From 20f1a77008548040cc22245230aaa24c547af178 Mon Sep 17 00:00:00 2001
From: Sanyog Kale <skale@habana.ai>
Date: Fri, 18 Jun 2021 08:57:40 +0300
Subject: [PATCH] [SW-33447] gaudi2: handle flr interrupt

The handling of FLR interrupt is different for gaudi2 compared to
gaudi. This patch implements gaudi2 specific FLR handling function
and implements FLR handling sequence recommended by HW team.

Change-Id: I3f3261e061dcc63a0ea2b069e924eba9fcb94ac8
---

diff --git a/arch/gaudi2/common.c b/arch/gaudi2/common.c
index 704a533..7af30f7 100644
--- a/arch/gaudi2/common.c
+++ b/arch/gaudi2/common.c
@@ -17,6 +17,8 @@
 #include <pcie_phy_masks.h>
 #include <pcie_dbi_regs.h>
 #include <pcie_dbi_masks.h>
+#include <cpu_if_regs.h>
+#include <cpu_if_masks.h>
 
 #include <cpu_timestamp_regs.h>
 #include <psoc_arc0_cfg_regs.h>
@@ -586,3 +588,17 @@
 {
 	return (u32)mmPCIE_DBI_BASE & 0xFFFFFFF;
 }
+
+void pci_dev_specific_flr(void)
+{
+	regmap_writel(1, mmPCIE_AUX_PF_FLR_ACTIVE);
+	regmap_writel(1, mmPCIE_AUX_PF_FLR_DONE);
+	regmap_writel(1, mmPCIE_AUX_FLR_INT);
+
+	/*
+	 * Clear PCIe SPI INTR. This is required to clear
+	 * FLR interrupt as recommended by HW team.
+	 */
+	regmap_writel(CPU_IF_PCIE_SPI_INTR_CLR_VAL_MASK,
+			mmCPU_IF_PCIE_SPI_INTR_CLR);
+}
diff --git a/arch/gaudi2_arc_gnu/common.c b/arch/gaudi2_arc_gnu/common.c
index f657bf9..7625cb5 100644
--- a/arch/gaudi2_arc_gnu/common.c
+++ b/arch/gaudi2_arc_gnu/common.c
@@ -21,6 +21,8 @@
 #include <cpu_timestamp_regs.h>
 #include <psoc_reset_conf_regs.h>
 #include <cpu_timestamp_regs.h>
+#include <cpu_if_regs.h>
+#include <cpu_if_masks.h>
 
 #include <stdint.h>
 #include <hl_boot_if.h>		/* requires stdint.h */
@@ -479,3 +481,17 @@
 	*loadaddr = copy_preboot();
 	return true;
 }
+
+void pci_dev_specific_flr(void)
+{
+	regmap_writel(1, mmPCIE_AUX_PF_FLR_ACTIVE);
+	regmap_writel(1, mmPCIE_AUX_PF_FLR_DONE);
+	regmap_writel(1, mmPCIE_AUX_FLR_INT);
+
+	/*
+	 * Clear PCIe SPI INTR. This is required to clear
+	 * FLR interrupt as recommended by HW team.
+	 */
+	regmap_writel(CPU_IF_PCIE_SPI_INTR_CLR_VAL_MASK,
+			mmCPU_IF_PCIE_SPI_INTR_CLR);
+}
diff --git a/preboot/module_pci.c b/preboot/module_pci.c
index f575741..2987e5e 100644
--- a/preboot/module_pci.c
+++ b/preboot/module_pci.c
@@ -54,6 +54,13 @@
 	regmap_writel(1, mmPCIE_AUX_LTSSM_EN);
 }
 
+__weak void pci_dev_specific_flr(void)
+{
+	regmap_writel(1, mmPCIE_AUX_PF_FLR_ACTIVE);
+	regmap_writel(1, mmPCIE_AUX_FLR_INT);
+	regmap_writel(1, mmPCIE_AUX_PF_FLR_DONE);
+}
+
 __weak void hl_pci_iatu_init(void) {}
 
 void module_pci_setup(void)
@@ -108,9 +115,7 @@
 	unsigned int rst_src;
 
 	if (pci_sts.flr_reg_val == 1) {
-		regmap_writel(1, mmPCIE_AUX_PF_FLR_ACTIVE);
-		regmap_writel(1, mmPCIE_AUX_FLR_INT);
-		regmap_writel(1, mmPCIE_AUX_PF_FLR_DONE);
+		pci_dev_specific_flr();
 		preboot_puts("FLR received\n");
 	}
 
